stratum/hal/lib/barefoot/bf_pipeline_utils_test.cc

```diff
@@ -0,0 +1,240 @@
+  0x6a, 0x35, 0x8d, 0xf5, 0x09, 0x86, 0x16, 0xc3, 0x03, 0xa8, 0x0e, 0x41,
+  0x19, 0xf2, 0x3c, 0x1e, 0x74, 0xa3, 0xc5, 0x5e, 0x8c, 0x9a, 0xe0, 0x71,
+  0x13, 0x32, 0x6e, 0xe2, 0x0c, 0x5a, 0x6c, 0xc6, 0x66, 0x75, 0x5e, 0xde,
+  0x20, 0xa3, 0xa9, 0x9a, 0x5b, 0x3d, 0xd3, 0xb1, 0xd5, 0x50, 0x09, 0xcf,
+  0x76, 0xdd, 0xd5, 0x14, 0x3c, 0x61, 0x19, 0x37, 0xb7, 0x6c, 0x7c, 0x2d,
+  0xa9, 0x78, 0x6e, 0xfb, 0xa2, 0x8b, 0xa6, 0x95, 0xfe, 0x8b, 0x19, 0xd5,
+  0x47, 0xd1, 0xb2, 0xe7, 0xd6, 0xa9, 0x3c, 0xb9, 0xe0, 0x10, 0x6a, 0x76,
+  0xe4, 0xb2, 0x9d, 0x7a, 0xa3, 0x67, 0xdf, 0x7d, 0xea, 0xfa, 0xd2, 0xf3,
+  0x01, 0x7d, 0x6e, 0xff, 0x25, 0x8b, 0xd4, 0x59, 0xa8, 0x77, 0xfb, 0x6c,
+  0xf4, 0x4a, 0x9f, 0xad, 0xfe, 0xe7, 0x2e, 0xcb, 0x12, 0x6e, 0x8a, 0xc3,
+  0xd5, 0xe4, 0x6d, 0xac, 0xfe, 0x93, 0x72, 0xdf, 0xaf, 0xff, 0x7c, 0xd7,
+  0x01, 0xfd, 0x9f, 0x02, 0x1a, 0x0a, 0xc6, 0xb3, 0xb2, 0x96, 0xde, 0xbc,
+  0x88, 0x69, 0xba, 0xdd, 0xc7, 0x8d, 0x0e, 0xb3, 0x68, 0xdd, 0x28, 0x93,
+  0x4a, 0xad, 0x26, 0xb3, 0xa4, 0x48, 0x98, 0x51, 0xc1, 0x9b, 0x94, 0xa7,
+  0x7b, 0x46, 0x4b, 0xf9, 0xa5, 0x6d, 0xa9, 0xe6, 0x9b, 0x28, 0xf3, 0xcd,
+  0x7f, 0xa5, 0xfb, 0x7f, 0xeb, 0xfe, 0x8f, 0xa5, 0x5c, 0x78, 0x70, 0xff,
+  0x9f, 0x30, 0xfe, 0x5f, 0xab, 0xa8, 0x1b, 0xdc, 0xff, 0xfb, 0xf5, 0xbf,
+  0x43, 0x30, 0xd4, 0xff, 0x13, 0xdd, 0xff, 0x43, 0x71, 0x40, 0x6b, 0x02,
+  0xb7, 0xff, 0xbb, 0xd6, 0xff, 0x73, 0xdd, 0x7f, 0x93, 0xff, 0x7f, 0xfd,
+  0xfa, 0x8f, 0xf8, 0xc4, 0x87, 0xfc, 0x9f, 0x82, 0x07, 0x19, 0x72, 0x23,
+  0xa2, 0x82, 0x3e, 0x42, 0x2e, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+  0x00, 0x00, 0x00, 0xc0, 0x5c, 0xf9, 0x00, 0xbe, 0x32, 0xc5, 0x34, 0x00,
+  0x28, 0x00, 0x00
+  };
+  const unsigned int my_pipe_tgz_len = 495; // generated using ... | wc
+  std::string my_pipe_tgz(my_pipe_tgz_bytes,
+      my_pipe_tgz_bytes + my_pipe_tgz_len / sizeof my_pipe_tgz_bytes[0]);
+  ASSERT_EQ(my_pipe_tgz_len, my_pipe_tgz.size());
+
+  ::p4::v1::ForwardingPipelineConfig p4_config;
+  p4_config.set_p4_device_config(my_pipe_tgz);
+  ASSERT_EQ(my_pipe_tgz_len, p4_config.p4_device_config().size());
+  *(p4_config.mutable_p4info()) = bf_config.programs()[0].p4info();
+
+  BfPipelineConfig extracted_bf_config;
+  EXPECT_OK(ExtractBfPipelineConfig(p4_config, &extracted_bf_config));
+  VLOG(1) << extracted_bf_config.DebugString();
+
+  EXPECT_TRUE(ProtoEqual(bf_config, extracted_bf_config));
+}
+
+TEST(ExtractBfPipelineTest, RandomBytes) {
+  ::p4::v1::ForwardingPipelineConfig p4_config;
+  p4_config.set_p4_device_config("<random vendor blob>");
+
+  BfPipelineConfig extracted_bf_config;
+  EXPECT_FALSE(ExtractBfPipelineConfig(p4_config, &extracted_bf_config).ok());
+}
+
+TEST(BfPipelineConvertTest, ToLegacyBfPiFormat) {
+  BfPipelineConfig bf_config;
+  ASSERT_OK(ParseProtoFromString(bf_config_1pipe_str, &bf_config));
+
+  std::string expected_config;
+  absl::CUnescape(
+      "\\x5\\0\\0\\0prog1\\x9\\0\\0\\0<raw bin>\\xc\\0\\0\\0{json: true}",
```

Comment:
What's the reason for using this encoding over `\xab\x00\xff`?

Reply:
string constructor was stopping at the first \0, but I think I can fix this with a length param

Reply:
fixed